src/router_core/modules/mobile_sync/mobile.c

```diff
@@ -0,0 +1,860 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "module.h"
+#include "router_core_private.h"
+#include "core_events.h"
+#include "route_control.h"
+#include <qpid/dispatch/ctools.h>
+#include <qpid/dispatch/log.h>
+#include <qpid/dispatch/compose.h>
+#include <qpid/dispatch/message.h>
+#include <stdio.h>
+#include <inttypes.h>
+
+#define PROTOCOL_VERSION 1
+static const char *OPCODE     = "opcode";
+static const char *MAR        = "MAR";
+static const char *MAU        = "MAU";
+static const char *ID         = "id";
+static const char *PV         = "pv";
+static const char *AREA       = "area";
+static const char *MOBILE_SEQ = "mobile_seq";
+static const char *HINTS      = "hints";
+static const char *ADD        = "add";
+static const char *DEL        = "del";
+static const char *EXIST      = "exist";
+static const char *HAVE_SEQ   = "have_seq";
+
+//
+// Address.sync_mask bit values
+//
+#define ADDR_SYNC_DELETION_WAS_BLOCKED  0x00000001
```

Comment:
Could this flag handling the blocking of the address deletion be better implemented by turning addr->block_deletion into a reference counter instead?

Reply:
Not a pointer reference counter, to be clearer.   A counter that prevents deletion when incremented?  Call qdr_check_addr_CT when decrementing causes it to hit zero?

Reply:
Thank you for this suggestion.  It has been incorporated in the latest squash/rebase.